Add synopsis and description for Zacas (Atomic CAS) extension (#516)

syedowaisalishah opened the pull request:
This PR adds the synopsis and description for the Zacas (Atomic CAS) extension instructions.
- Updated amocas.d.yaml, amocas.q.yaml, amocas.w.yaml.
- Extracted details from the RISC-V Unprivileged ISA Spec.

issue #516.

syedowaisalishah commented:
@ThinkOpenly Thanks for the detailed feedback! I'll refine the commit message and update the descriptions to align with the ISA spec. Also, I'll fix the operation(): | formatting issue and push the changes shortly.

dhower-qc commented on arch/inst/Zacas/amocas.d.yaml:
```diff
@@ -28,3 +64,7 @@ access:
   vu: always
 data_independent_timing: false
 operation(): |
+  temp = mem[X(rs1)]
```
We'll need to work on the IDL here. A few things to note:

- the address needs to go through translation (what is in X[rs1] is a virtual address)
- the swap must be atomic, which is going to require a call to what IDL calls a "builtin" function.
- Needs to handle the differences in RV32/64
- Needs to check for x0
- Needs to check all the exception cases

My recommendation is to just leave this empty for now, and we will implement it through #517

dhower-qc commented on arch/inst/Zacas/amocas.d.yaml:
```diff
@@ -3,9 +3,45 @@
+
+  **Additional Considerations:**
+  - The first register in a pair must be even-numbered; encodings with
+    odd-numbered registers in rd or rs2 are reserved.
```
We don't have the even/odd requirements reflected in the YAML. To do that:

- We need to split the encoding for rv32/rv64, since they are different
- For the rv32 version, need to add `not:` and cover all the odd register numbers

example:

```yaml
encoding:
  RV32:
    match: ...
    variables:
      - name: rd
        location: 11-7
        not: [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31]
      ..
  RV64:
    match: ...
    # ...
```

syedowaisalishah commented:
@dhower-qc Thanks for the feedback! I'll update the YAML encoding for even/odd constraints, leave operation(): | empty as suggested, and check the failing tests. Will push updates soon.

syedowaisalishah commented:
@ThinkOpenly and @dhower-qc , I've updated the YAML encoding for even/odd constraints, left operation(): | empty as suggested, and reviewed exception handling. Let me know if any further refinements are needed.

ThinkOpenly commented on arch/inst/Zacas/amocas.d.yaml:
```diff
@@ -3,9 +3,24 @@
+  - Compares with rd.
+  - If equal, stores the 64-bit value from rs2 to rs1.
+  - Updates rd with the loaded value.
+
```
This comment is still valid. Try to use the ratified text from the RISC-V ISA Specification when possible. I cut and pasted the above material, and only edited it for specificity (since this text is not associated with all of AMOCAS.W/D/Q, for example).

syedowaisalishah commented:
@ThinkOpenly and @dhower-qc I've updated the encoding to separate RV32 and RV64 constraints as suggested. I also incorporated the ratified text from the RISC-V ISA Specification for amocas.d,q and w.

ThinkOpenly commented on arch/inst/Zacas/amocas.d.yaml:
```diff
@@ -1,26 +1,68 @@
+  comparison is bitwise equal, then stores the 64-bit value held in rs2 to the
+  original address in rs1. The value loaded from memory is placed into
+  register rd.
+
```
Let's insert the intermediate text which also refers to "AMOCAS.D" here:
> Just as for AMOs in the A extension, AMOCAS.D requires that the address held in rs1 be naturally aligned to the size of the operand (i.e., eight-byte aligned for doublewords). And the same exception options apply if the address is not naturally aligned.
> 
> Just as for AMOs in the A extension, the AMOCAS.D optionally provides release consistency semantics, using the aq and rl bits, to help implement multiprocessor synchronization. The memory operation performed by an AMOCAS.D, when successful, has acquire semantics if aq bit is 1 and has release semantics if rl bit is 1. The memory operation performed by an AMOCAS.W/D/Q, when not successful, has acquire semantics if aq bit is 1 but does not have release semantics, regardless of rl.

dhower-qc reviewed (approved):
👍
